File: projects/mcp-8004-enterprise/agents_implementation/launch-bridges.js
/**
 * launch-bridges.js
 *
 * Spawns all onboarding oracle bridge processes.
 *
 * Usage:
 *   node agents_implementation/launch-bridges.js \
 *     --onboarding-registry 0x... \
 *     --aml-contract        0x... \
 *     --credit-contract     0x... \
 *     --legal-contract      0x... \
 *     --setup-contract      0x... \
 *     --rpc  http://127.0.0.1:8545 \
 *     --privkey 0x...
 *
 * All flags also readable from env (see individual bridge scripts for ENV names).
 * Optional governance flags: --flow-auth, --reputation-gate, --autonomy-bounds,
 *                            --action-permit, --identity-registry
 *
 * Ctrl-C stops all child processes.
 */

import { spawn } from 'node:child_process';
import path      from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// ── CLI / env helper ──────────────────────────────────────────────────────────

function arg(flag, envVar) {
  const idx = process.argv.indexOf(flag);
  return (idx !== -1 && process.argv[idx + 1]) ? process.argv[idx + 1] : process.env[envVar];
}

// ── Common flags (forwarded to all bridges) ───────────────────────────────────

const RPC           = arg('--rpc',                   'RPC_URL')                   ?? 'http://127.0.0.1:8545';
const PRIVKEY       = arg('--privkey',               'ORACLE_PRIVATE_KEY');
const SIGNER_TYPE   = arg('--signer-type',           'SIGNER_TYPE')               ?? 'local';
const VAULT_URL     = arg('--vault-url',             'VAULT_URL');
const VAULT_ADDR    = arg('--vault-address',         'ORACLE_ADDRESS');
const FLOW_AUTH     = arg('--flow-auth',             'FLOW_AUTH_ADDRESS');
const REP_GATE      = arg('--reputation-gate',       'REPUTATION_GATE_ADDRESS');
const AUTO_BOUNDS   = arg('--autonomy-bounds',       'AUTONOMY_BOUNDS_ADDRESS');
const ACTION_PERMIT = arg('--action-permit',         'ACTION_PERMIT_ADDRESS');
const IDENTITY_REG  = arg('--identity-registry',     'IDENTITY_REGISTRY_ADDRESS');

// ── Contract addresses ────────────────────────────────────────────────────────

const ONBOARDING_REG  = arg('--onboarding-registry', 'ONBOARDING_REGISTRY_ADDRESS');
const AML_CONTRACT    = arg('--aml-contract',         'AML_CONTRACT_ADDRESS');
const CREDIT_CONTRACT = arg('--credit-contract',      'CREDIT_CONTRACT_ADDRESS');
const LEGAL_CONTRACT  = arg('--legal-contract',       'LEGAL_CONTRACT_ADDRESS');
const SETUP_CONTRACT  = arg('--setup-contract',       'SETUP_CONTRACT_ADDRESS');

// ── Bounds monitor ────────────────────────────────────────────────────────────

const BOUNDS_PORT    = arg('--bounds-port') ?? '9090';
const BOUNDS_MOCK    = process.argv.includes('--bounds-mock');

// ── Agent IDs ─────────────────────────────────────────────────────────────────

const AML_AGENT_ID      = arg('--aml-agent-id',      'AML_AGENT_ID')      ?? '0';
const CREDIT_AGENT_ID   = arg('--credit-agent-id',   'CREDIT_AGENT_ID')   ?? '1';
const LEGAL_AGENT_ID    = arg('--legal-agent-id',    'LEGAL_AGENT_ID')    ?? '3';
const ENTITY_AGENT_ID   = arg('--entity-agent-id',   'ENTITY_AGENT_ID')   ?? '4';
const ACCOUNT_AGENT_ID  = arg('--account-agent-id',  'ACCOUNT_AGENT_ID')  ?? '5';
const PRODUCT_AGENT_ID  = arg('--product-agent-id',  'PRODUCT_AGENT_ID')  ?? '6';
const HF_DOC_AGENT_ID   = arg('--hf-doc-agent-id',   'HF_DOC_AGENT_ID')   ?? '7';
const HF_CREDIT_AGENT_ID= arg('--hf-credit-agent-id','HF_CREDIT_AGENT_ID')  ?? '8';
const HF_LEGAL_AGENT_ID = arg('--hf-legal-agent-id', 'HF_LEGAL_AGENT_ID') ?? '9';

// ── Governance optional flags ─────────────────────────────────────────────────

function govFlags() {
  const flags = [];
  if (FLOW_AUTH)     { flags.push('--flow-auth',         FLOW_AUTH); }
  if (REP_GATE)      { flags.push('--reputation-gate',   REP_GATE); }
  if (AUTO_BOUNDS)   { flags.push('--autonomy-bounds',   AUTO_BOUNDS); }
  if (ACTION_PERMIT) { flags.push('--action-permit',     ACTION_PERMIT); }
  if (IDENTITY_REG)  { flags.push('--identity-registry', IDENTITY_REG); }
  if (SIGNER_TYPE !== 'local') { flags.push('--signer-type', SIGNER_TYPE); }
  if (VAULT_URL)     { flags.push('--vault-url',         VAULT_URL); }
  if (VAULT_ADDR)    { flags.push('--vault-address',     VAULT_ADDR); }
  return flags;
}

function signerFlags() {
  const flags = ['--rpc', RPC];
  if (PRIVKEY) flags.push('--privkey', PRIVKEY);
  return [...flags, ...govFlags()];
}

// ── Bridge definitions ────────────────────────────────────────────────────────

const bridges = [
  // ── Bounds monitor (always started; --mock skips on-chain calls) ───────────
  {
    name: 'bounds-monitor',
    script: 'bounds-monitor.js',
    args: [
      '--port', BOUNDS_PORT,
      '--rpc',  RPC,
      ...(PRIVKEY      ? ['--privkey',        PRIVKEY]    : []),
      ...(AUTO_BOUNDS  ? ['--autonomy-bounds', AUTO_BOUNDS]: []),
      ...(BOUNDS_MOCK || !AUTO_BOUNDS ? ['--mock'] : []),
    ],
  },
  // ── Bank side ──────────────────────────────────────────────────────────────
  AML_CONTRACT && {
    name: 'aml-bridge',
    script: 'aml-bridge.js',
    args: ['--contract', AML_CONTRACT, '--agent-id', AML_AGENT_ID, ...signerFlags()],
  },
  CREDIT_CONTRACT && {
    name: 'credit-risk-bridge',
    script: 'credit-risk-bridge.js',
    args: ['--contract', CREDIT_CONTRACT, '--agent-id', CREDIT_AGENT_ID, ...signerFlags()],
  },
  LEGAL_CONTRACT && {
    name: 'legal-bridge',
    script: 'legal-bridge.js',
    args: ['--contract', LEGAL_CONTRACT, '--agent-id', LEGAL_AGENT_ID, ...signerFlags()],
  },
  (ONBOARDING_REG && AML_CONTRACT && CREDIT_CONTRACT && LEGAL_CONTRACT) && {
    name: 'orchestrator-bridge',
    script: 'onboarding-orchestrator-bridge.js',
    args: [
      '--onboarding-registry', ONBOARDING_REG,
      '--aml-contract',        AML_CONTRACT,
      '--credit-contract',     CREDIT_CONTRACT,
      '--legal-contract',      LEGAL_CONTRACT,
      ...signerFlags(),
    ],
  },
  (ONBOARDING_REG && SETUP_CONTRACT) && {
    name: 'client-setup-bridge',
    script: 'client-setup-bridge.js',
    args: [
      '--onboarding-registry', ONBOARDING_REG,
      '--setup-contract',      SETUP_CONTRACT,
      '--entity-agent-id',     ENTITY_AGENT_ID,
      '--account-agent-id',    ACCOUNT_AGENT_ID,
      '--product-agent-id',    PRODUCT_AGENT_ID,
      ...signerFlags(),
    ],
  },
  // ── Hedge fund side ────────────────────────────────────────────────────────
  (AML_CONTRACT && CREDIT_CONTRACT) && {
    name: 'hf-document-bridge',
    script: 'hf-document-bridge.js',
    args: [
      '--aml-contract',    AML_CONTRACT,
      '--credit-contract', CREDIT_CONTRACT,
      '--agent-id',        HF_DOC_AGENT_ID,
      ...signerFlags(),
    ],
  },
  CREDIT_CONTRACT && {
    name: 'hf-credit-negotiator-bridge',
    script: 'hf-credit-negotiator-bridge.js',
    args: ['--credit-contract', CREDIT_CONTRACT, '--agent-id', HF_CREDIT_AGENT_ID, ...signerFlags()],
  },
  LEGAL_CONTRACT && {
    name: 'hf-legal-bridge',
    script: 'hf-legal-bridge.js',
    args: ['--legal-contract', LEGAL_CONTRACT, '--agent-id', HF_LEGAL_AGENT_ID, ...signerFlags()],
  },
].filter(Boolean);

// ── Spawn ─────────────────────────────────────────────────────────────────────

const children = [];

for (const { name, script, args } of bridges) {
  const scriptPath = path.join(__dirname, script);
  console.log(`[launch-bridges] Starting ${name}`);

  const child = spawn(process.execPath, [scriptPath, ...args], {
    stdio: 'inherit',
    env: process.env,
  });

  child.on('exit', (code, signal) =>
    console.log(`[launch-bridges] ${name} exited (code=${code} signal=${signal})`)
  );

  children.push({ name, child });
}

console.log(`\n[launch-bridges] ${children.length} bridge(s) running. Ctrl-C to stop all.\n`);

function shutdown() {
  console.log('\n[launch-bridges] Shutting down…');
  for (const { name, child } of children) {
    console.log(`[launch-bridges]   stopping ${name}`);
    child.kill('SIGTERM');
  }
  process.exit(0);
}

process.on('SIGINT',  shutdown);
process.on('SIGTERM', shutdown);

File: projects/mcp-8004-enterprise/agents_implementation_py/launch_bridges.py
"""
launch_bridges.py — Launch all Python oracle bridges as subprocesses.
Python counterpart to launch-bridges.js.

Pass the same flags as the Node.js version:
    python launch_bridges.py \
        --rpc                 http://127.0.0.1:8545 \
        --privkey             0x<YOUR_KEY> \
        --onboarding-registry 0x... \
        --aml-contract        0x... \
        --credit-contract     0x... \
        --legal-contract      0x... \
        --setup-contract      0x... \
        [--identity-registry  0x...] \
        [--flow-auth          0x...] \
        [--reputation-gate    0x...] \
        [--autonomy-bounds    0x...] \
        [--action-permit      0x...] \
        [--aml-agent-id       0]     \
        [--credit-agent-id    1]     \
        [--legal-agent-id     2]     \
        [--entity-agent-id    4]     \
        [--account-agent-id   5]     \
        [--product-agent-id   6]     \
        [--hf-doc-agent-id    7]     \
        [--hf-credit-agent-id 8]     \
        [--hf-legal-agent-id  9]
"""

from __future__ import annotations

import argparse
import signal
import subprocess
import sys
from pathlib import Path

HERE = Path(__file__).parent

# (module, extra_flags_fn)
# Each tuple maps to a Python bridge module and a function that returns
# the extra flags specific to that bridge (given the parsed args namespace).
BRIDGES = [
    (
        "bridges.aml_bridge",
        lambda a: ["--contract", a.aml_contract, "--agent-id", a.aml_agent_id],
    ),
    (
        "bridges.credit_risk_bridge",
        lambda a: ["--contract", a.credit_contract, "--agent-id", a.credit_agent_id],
    ),
    (
        "bridges.legal_bridge",
        lambda a: ["--contract", a.legal_contract, "--agent-id", a.legal_agent_id],
    ),
    (
        "bridges.client_setup_bridge",
        lambda a: [
            "--onboarding-registry", a.onboarding_registry,
            "--setup-contract",      _require(a.setup_contract, "--setup-contract"),
            "--entity-agent-id",     a.entity_agent_id,
            "--account-agent-id",    a.account_agent_id,
            "--product-agent-id",    a.product_agent_id,
        ],
    ),
    (
        "bridges.hf_document_bridge",
        lambda a: [
            "--aml-contract",    a.aml_contract,
            "--credit-contract", a.credit_contract,
            "--agent-id",        a.hf_doc_agent_id,
        ],
    ),
    (
        "bridges.hf_credit_negotiator_bridge",
        lambda a: ["--credit-contract", a.credit_contract, "--agent-id", a.hf_credit_agent_id],
    ),
    (
        "bridges.hf_legal_bridge",
        lambda a: ["--legal-contract", a.legal_contract, "--agent-id", a.hf_legal_agent_id],
    ),
    (
        "bridges.onboarding_orchestrator_bridge",
        lambda a: [
            "--onboarding-registry", a.onboarding_registry,
            "--aml-contract",        a.aml_contract,
            "--credit-contract",     a.credit_contract,
            "--legal-contract",      a.legal_contract,
        ],
    ),
]

# ── Common flags injected into every bridge process ───────────────────────────
COMMON_FLAGS = [
    ("--rpc",               "rpc"),
    ("--privkey",           "privkey"),
    ("--signer-type",       "signer_type"),
    ("--vault-url",         "vault_url"),
    ("--vault-address",     "vault_address"),
    ("--identity-registry", "identity_registry"),
    ("--flow-auth",         "flow_auth"),
    ("--reputation-gate",   "reputation_gate"),
    ("--autonomy-bounds",   "autonomy_bounds"),
    ("--action-permit",     "action_permit"),
]


def _require(val: str | None, flag: str) -> str:
    if val is None:
        raise ValueError(f"{flag} not provided")
    return val


def _build_common(args: argparse.Namespace) -> list[str]:
    result = []
    for flag, attr in COMMON_FLAGS:
        val = getattr(args, attr, None)
        if val:
            result.extend([flag, val])
    return result


def main() -> None:
    p = argparse.ArgumentParser(description="Launch all Python oracle bridges")
    p.add_argument("--rpc",               default="http://127.0.0.1:8545")
    p.add_argument("--privkey",           default=None)
    p.add_argument("--signer-type",       dest="signer_type", default="local")
    p.add_argument("--vault-url",         dest="vault_url",   default=None)
    p.add_argument("--vault-address",     dest="vault_address", default=None)
    p.add_argument("--onboarding-registry", dest="onboarding_registry", required=True)
    p.add_argument("--aml-contract",      dest="aml_contract",    required=True)
    p.add_argument("--credit-contract",   dest="credit_contract", required=True)
    p.add_argument("--legal-contract",    dest="legal_contract",  required=True)
    p.add_argument("--setup-contract",    dest="setup_contract",  default=None)
    p.add_argument("--identity-registry", dest="identity_registry", default=None)
    p.add_argument("--flow-auth",         dest="flow_auth",         default=None)
    p.add_argument("--reputation-gate",   dest="reputation_gate",   default=None)
    p.add_argument("--autonomy-bounds",   dest="autonomy_bounds",   default=None)
    p.add_argument("--action-permit",     dest="action_permit",     default=None)
    p.add_argument("--aml-agent-id",      dest="aml_agent_id",      default="0")
    p.add_argument("--credit-agent-id",   dest="credit_agent_id",   default="1")
    p.add_argument("--legal-agent-id",    dest="legal_agent_id",    default="2")
    p.add_argument("--entity-agent-id",   dest="entity_agent_id",   default="4")
    p.add_argument("--account-agent-id",  dest="account_agent_id",  default="5")
    p.add_argument("--product-agent-id",  dest="product_agent_id",  default="6")
    p.add_argument("--hf-doc-agent-id",   dest="hf_doc_agent_id",   default="7")
    p.add_argument("--hf-credit-agent-id",dest="hf_credit_agent_id",default="8")
    p.add_argument("--hf-legal-agent-id", dest="hf_legal_agent_id", default="9")
    args = p.parse_args()

    common = _build_common(args)
    procs: list[subprocess.Popen] = []

    for module, extra_fn in BRIDGES:
        try:
            extra = extra_fn(args)
        except Exception as exc:
            print(f"Skipping {module}: {exc}")
            continue
        # Filter out None values
        extra = [str(v) for v in extra if v is not None]
        cmd = [sys.executable, "-m", module] + common + extra
        print(f"Starting {module}…")
        proc = subprocess.Popen(cmd, cwd=HERE)
        procs.append(proc)

    print(f"\n{len(procs)} Python bridges started. Press Ctrl-C to stop all.\n")

    def _shutdown(sig, frame):
        print("\nShutting down all bridges…")
        for proc in procs:
            proc.terminate()
        for proc in procs:
            proc.wait()
        sys.exit(0)

    signal.signal(signal.SIGINT,  _shutdown)
    signal.signal(signal.SIGTERM, _shutdown)

    for proc in procs:
        proc.wait()


if __name__ == "__main__":
    main()
